Add mount_facts module

s-hertel opened the pull request:
##### SUMMARY

Add a module to gather mount information, including devices that don't look like paths (which are excluded by default fact gathering). Renamed to `mount_facts` for less typing, since the module is more generic now.

Fixes #24644

Marked as draft until I've added:
* [x] Tests
* [x] Changelog

##### ISSUE TYPE

- Feature Pull Request

##### ADDITIONAL INFORMATION

<!--- Include additional information to help people understand the change here -->

<!--- Paste verbatim command output below, e.g. before and after your change -->

```paste below

```

s-hertel committed:
Add a mount_facts module capable of gathering mounts skipped by default
fact gathering

Add a sources option to select the origin of the mount facts, defaulting
to standard locations (/proc/mounts, /etc/mtab, /etc/fstab, /etc/vfstab,
/etc/mnttab/ /etc/filesystems (AIX-specific), and the output from the
mount binary if not other dynamic sources were found (as can be expected
for BSD or AIX).

Add options to filter on fstype and devices (supporting UNIX shell
wildcards).

Default to wait indefinitely to match the expectation of running as a
standalone task, but support configuring a timeout and timeout behavior
to support using the module as a default facts module.

Add unit tests

Add integration test, including setting up NFS mount

Changelog

s-hertel commented on lib/ansible/modules/remote_mount_facts.py:
```diff
@@ -0,0 +1,290 @@
+      - nfs
+      - nfs4
+      - vboxsf
+      - zfs
```
Since anyone using this in gather_facts will need to opt-in anyway, would it be better to make the default the full list of choices for convenience?

Should I add tmpfs to the list of choices?

bcoca commented:
bikeshed: extended_mount_facts

bcoca commented on lib/ansible/modules/remote_mount_facts.py:
```diff
@@ -0,0 +1,290 @@
+      - zfs
+  timeout:
+    description: The maximum number of seconds to query for each mount type in O(collect).
+    default: 20
```
maybe 0/none should be the default, most users of this module should not have too many issues and this should only be set if they do encounter them.

s-hertel replied:
I made the module default 0, which internally times out after .01 seconds.

bcoca replied:
why i was thinking None was better default, only pass/call timeout when `is not None`, can do same with 0 but I think `None` makes a better sentinel here.

s-hertel replied:
Only calling timeout when timeout `is not None/0` allows a hang, which means anyone enabling this module for gather_facts will also need to set the timeout. I think it should error after a short period of time by default instead (which is what I've done).

bcoca replied:
ok, i was just thinking most users won't have hangs but can have long waits for network items, since this is an explicit choice we could default to waiting, unlike 'default fact gathering' in which waiting can become too cumbersome.

s-hertel replied:
I guess I misunderstood the request. I thought 20 seconds was erring on an unreasonable amount of time to wait since it's per mount, but could make it longer too ¯\_(ツ)_/¯ Making `timeout(seconds)` conditional introduces what I'd consider to be a bug.

s-hertel replied:
>  ... unlike 'default fact gathering' in which waiting can become too cumbersome.

I am specifically thinking of people who are going to enable this as a FACTS_MODULE.

s-hertel replied:
I circled back around to this, and fixed using no timeout, thanks for pointing it out.

bcoca commented on lib/ansible/modules/remote_mount_facts.py:
```diff
@@ -0,0 +1,290 @@
+    description: The maximum number of seconds to query for each mount type in O(collect).
+    default: 20
+    type: int
+  timeout_handler:
```
bike shed should be orange! on_timeout?

s-hertel replied:
+1, renamed

bcoca commented on lib/ansible/modules/remote_mount_facts.py:
```diff
@@ -0,0 +1,290 @@
+    type: list
+    elements: str
+    required: true
+    choices:
```
maybe instead of an enum a filters/patterns so even if new mount types appear user can use module w/o code update?

bcoca replied:
i.e i have vfat from my usb drive, but also hfs from a friend's mac drive i was 'rescuing', trying to keep up with the list will can be daunting

s-hertel replied:
I replaced this option with `devices` and `fstypes` to filter using fnmatch. This makes the module much easier to test too. Let me know what you think.

bcoca commented on lib/ansible/modules/remote_mount_facts.py:
```diff
@@ -0,0 +1,290 @@
+
+    mnttab = False
+    mtab_entries = ""
+    if os.path.exists("/etc/mtab"):
```
no need for 'ifs', since get_file_content already checks:
```
    if os.path.exists(path) and os.access(path, os.R_OK):
```

s-hertel replied:
good spotting, refactored this function

bcoca commented on lib/ansible/modules/remote_mount_facts.py:
```diff
@@ -0,0 +1,290 @@
+    mtab_entries = ""
+    if os.path.exists("/etc/mtab"):
+        mtab_entries = get_file_content("/etc/mtab", "")
+    if not mtab_entries and os.path.exists("/proc/mounts"):
```
same here, just `if not mtab_entries:`

bcoca replied:
or just a for loop on the files and escape once we have mtab entries?

s-hertel replied:
Fixed

bcoca commented on lib/ansible/modules/remote_mount_facts.py:
```diff
@@ -0,0 +1,290 @@
+    mount_output = ""
+    if mount_path:
+        rc, mount_out, err = module.run_command(mount_path)
+
```
check rc/err to warn user in case calling mount fails

s-hertel replied:
fixed

bcoca commented on lib/ansible/modules/extended_mount_facts.py:
```diff
@@ -0,0 +1,282 @@
+    """
+    Parse a line of mount information and return a dictionary with mount details.
+    """
+    match = BSD_MOUNT_RE.match(line) or AIX_MOUNT_RE.match(line)
```
should we check OS/platform first in case some weird fs driver can match 'other platforms'?

s-hertel replied:
I added a check for sys.platform, but made it kind of lenient (allowing 'linux' to match) since it seems there is no standardization.

bcoca replied:
sorry, it was a bad idea on my part, it would need to include the distro lib (we do vendor) and a lot of complicated and fallback heavy querying. I forgot that 'platform detection' is a mess.

bcoca commented on lib/ansible/modules/extended_mount_facts.py:
```diff
@@ -0,0 +1,282 @@
+      - vboxsf
+      - zfs
+  timeout:
+    description: The maximum number of seconds to query for each mount type in O(collect).
```
for each mount, not mount type

s-hertel replied:
good spotting, wrote the docs before I wrote the code
```suggestion
    description: The maximum number of seconds to query for each mount.
```

bcoca commented on lib/ansible/modules/extended_mount_facts.py:
```diff
@@ -0,0 +1,282 @@
+    """
+    mount_path = module.get_bin_path("mount")
+    mount_output = ""
+    if mount_path:
```
we might want to add timeout here too as calling mount when there is a bad/unresponsive mount point can also hang for a long time

s-hertel replied:
The timeout now applies to executing mount too.

Akasurde commented on lib/ansible/modules/extended_mount_facts.py:
```diff
@@ -0,0 +1,298 @@
+    """
+    mount_path = module.get_bin_path("mount")
+    mount_output = ""
+    if mount_path:
```
Let's bail out early ... 

```
mounts = []
if mount_path is None:
   return mounts
```

s-hertel replied:
I wanted to have 1 return statement, but yeah I just added a second...

webknjaz commented on lib/ansible/modules/extended_mount_facts.py:
```diff
@@ -0,0 +1,300 @@
+    device: str,
+    uuids: dict[str, str],
+    udevadm_uuid: Callable[[str], str]
+) -> tuple[Optional[dict], Optional[str]]:
```
Nit: `Optional` is considered confusing, so many linters recommend being explicit about its actual semantics:
```suggestion
) -> tuple[dict | None, str | None]:
```

s-hertel replied:
Thanks, I forgot about that.

s-hertel replied:
Fixed

s-hertel committed:
flatten return data and fix RETURN docs

return the mount points as keys, to distinguish from the AIX mount
attribute "mount" (not the mount point) which could be in the value

since AIX doesn't shadow fstype or devices, keep adding those as
part of the value for filtering on the normalized value

use generators and update the shape of the test data

yield the source and line(s) corresponding to a mount for easier
debugging, return under the key "ansible_context"

s-hertel committed:
Remove unused variable

Remove redundant "or ''"

Simplify AIX mount regex

s-hertel committed:
refactor for readability

s-hertel committed:
Change defaults to None

Or not, change default to empty list (None default doesn't work with type list results in [None])

Change timeout to a float

Rework description for sources

s-hertel committed:
remove _ prefix since it's redundant

s-hertel committed:
typos

s-hertel committed:
fix defaults as null

s-hertel commented on lib/ansible/modules/mount_facts.py:
```diff
@@ -0,0 +1,472 @@
+    """Helper returning the argument spec."""
+    return dict(
+        sources=dict(type="list", elements="str", default=None),
+        mount_binary=dict(default="mount", type="raw"),
```
Hopefully I've captured this pretty well in the documentation, but wanted to add a note on the implementation, since type `raw` is out of the ordinary. This gets used implicitly as a dynamic source only for hosts that don't have a file-based alternative, which is why it's configurable alongside sources. This is type `raw` to allow `mount_binary: ~` (to opt out of the fallback) since `~` is not a valid choice for a str/path type unless it is the default value. The default isn't `~` to allow BSD and AIX to collect the current mounts by default.

The mount_binary can also be configured explicitly in sources so it always runs, but this would be redundant for Linux (only really makes sense as the only dynamic source in that case, but the mount output is non-standardized, and parsing it is best-effort, so using a file-based source is superior).

mattclay commented on lib/ansible/modules/mount_facts.py:
```diff
@@ -0,0 +1,472 @@
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.facts import timeout as _timeout
+from ansible.module_utils.facts.hardware.linux import LinuxHardware, _replace_octal_escapes
```
It would be good to minimize the dependencies on (and changes to) the existing module_utils facts code. Instead of importing `_replace_octal_escapes`, a simple one-line function will suffice:

```python
def _replace_octal_escapes(value: str) -> str:
    return re.sub(r'(\\[0-7]{3})', lambda m: codecs.decode(m.group(0), 'unicode_escape'), value)
```

s-hertel replied:
Good point, fixed. I removed the dependency on LinuxHardware for the UUIDs too.

mattclay commented on lib/ansible/module_utils/facts/timeout.py:
```diff
@@ -48,7 +48,7 @@ def wrapper(*args, **kwargs):
             except multiprocessing.TimeoutError:
                 # This is an ansible.module_utils.common.facts.timeout.TimeoutError
-                raise TimeoutError('Timer expired after %s seconds' % timeout_value)
+                raise TimeoutError(f'{error_message} after {timeout_value} seconds')
```
Since this is a bug fix, it should go into a separate PR. That will allow it to be backported.

That will also have the benefit of removing the last change to module_utils from this PR (taking into account my other comment on `_replace_octal_escapes`).

s-hertel replied:
Fixed, I have opened this as https://github.com/ansible/ansible/pull/83945.

mattclay commented on test/units/modules/test_mount_facts.py:
```diff
@@ -0,0 +1,284 @@
+    fail_json as _fail_json,
+    set_module_args as _set_module_args
+)
+from units.modules._mount_facts_data import _aix7_2, _freebsd14_1, _openbsd6_4, _rhel9_4, _solaris11_2
```
Since none of our unit test code is user facing, there's no need to use `_` imports to mark anything private.

s-hertel replied:
Fixed

mattclay commented on test/integration/targets/mount_facts/aliases:
```diff
@@ -0,0 +1,4 @@
+shippable/posix/group1
+context/target
+needs/root
+skip/docker
```
This test should be marked `destructive` since it makes changes to the system.

s-hertel replied:
Thanks, good catch. I had written this test non-destructive initially.

nitzmahone commented on lib/ansible/modules/mount_facts.py:
```diff
@@ -0,0 +1,472 @@
+        mount_info = {}
+        for line in stanza:
+            attr, value = line.split("=", 1)
+            mount_info[attr.strip().rstrip()] = value.strip().rstrip()
```
`strip()` already works both ends of the string, so all the chained `rstrip()`s are seemingly superfluous.

s-hertel replied:
Fixed

nitzmahone reviewed (commented):
UX looks pretty solid and featureful. I assume we're going for rough compatibility with the existing mount facts, hence not providing further processing of some composite or conditionally-available values (e.g. `options: "rw,nosuid..."`, `uuid: N/A`)?

None of the following should block merge, but there are a number of things that could improve maintainability and robustness of the module implementation and tests:
* Define a few self-parsing dataclasses for the various input types; `-> t.Iterator[LinuxMountInfo]` is a lot more grok-able than `-> Generator[tuple[str, str, dict[str, str | int]]]`, and will also be self-documenting and fully type-checkable, while still being highly interoperable with `dict` (in and out).
* Avoid importing the existing facts impls, even if it means refactoring necessary parts into a new internal shared module_util- can add a lot to the module payload for little benefit.
* Avoid repeated inline heuristic determination of mount "shapes"- if we know we're Linux/BSD/AIX/whatever, only parse the relevant style(s). It seems like only a matter of time before someone tries to add something that will cause the wrong regex to fire (assuming it's not already possible).

Happy to work with you on any/all of these for clarification/ideas/collab impl/whatever.

s-hertel committed:
Revert changes to module_utils

Define LinuxHardware methods in mount_facts instead of relying on the default fact-gathering code

s-hertel committed:
Simplify return types with a dataclass

s-hertel committed:
Remove redundant rstrip()

s-hertel committed:
Make mount output matching more strict

s-hertel committed:
Fix flawed return type, return a list of dictionaries like default mount facts

the same mount point may have multiple devices, so returning a dictionary using the mount points as keys omits device information

s-hertel committed:
Adjust AIX filesystems return type since the mount attribute conflicts with the mount point field

s-hertel committed:
Fix test alias

s-hertel committed:
Add a better integration tests for mount fact order and UUIDs

s-hertel committed:
Update unit tests and add coverage for new functions

Remove redundant _ prefix

s-hertel commented:
> I assume we're going for rough compatibility with the existing mount facts, hence not providing further processing of some composite or conditionally-available values (e.g. options: "rw,nosuid...", uuid: N/A)?

@nitzmahone I did leave `options` as the raw string for compatibility. I was thinking a new key could be added for parsed options. On a related note, I ended up adding a dictionary of options for AIX /etc/filesystems fields (under the key "attributes" since "options" is taken), because one of the AIX attributes is "mount" (which would clobber the mount point). I'm wondering if I should change that to a list, if it would be more compatible for Linux mount options (without a corresponding =value).

I could omit `uuid` instead of setting it to `N/A`.  The format is closer to default fact-gathering now (I had overlooked that some mount points have multiple devices, so I changed it to a list of dictionaries), but it's not identical. The options get munged here https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/facts/hardware/linux.py#L600, and I left that out because it seemed like if I was using the options to create or modify mounts, the munged value would be an inconvenience.

I took at stab at the points you brought up around maintainability :crossed_fingers: There's probably more to do, but please let me know how it looks to you. I added a couple dataclasses to simplify type hints, removed the most of the reliance on module_utils, and tried to tighten up the regex matching by ensuring all lines match excluding the ones I know don't, instead of checking if any match.

s-hertel committed:
fix rebase fail, delete renamed file

s-hertel committed:
Fix type hint

s-hertel committed:
Prefer first mount from first source and reorder the default sources
to dynamic, static so the current mount still wins.

Add toggle "include_aggregate_mounts" for return aggregate_mounts in
addition to mount_points.

When "include_aggregate_mounts" is None (the default) and a mount point
is found in an individual source multiple times, a warning is given that
only the first is represented in mount_points and that configuring
include_aggregate_mounts (as True or False) will turn off the warning.

Change uuid default to None

Update the tests

nitzmahone commented on lib/ansible/modules/mount_facts.py:
```diff
@@ -0,0 +1,637 @@
+    description:
+      - A list of sources used to determine the mounts. Missing file sources (or empty files) are skipped.
+      - The C(mount_points) return value contains the first definition found for a mount point.
+      - Additional mounts to the same mount point are available from  C(aggregate_mounts) (if enabled).
```
```suggestion
      - Additional mounts to the same mount point are available from C(aggregate_mounts) (if enabled).
```
(remove extra space)

s-hertel replied:
Thanks, good catch.

nitzmahone commented on lib/ansible/modules/mount_facts.py:
```diff
@@ -0,0 +1,637 @@
+                    seen.add(mount_point)
+                else:
+                    # Warn about duplicate mounts within the same source since there may be multiple devices mounted.
+                    module.warn(f"Ignoring repeat mounts for {mount_point}. "
```
It'd be good to include which module sourced the warning (especially when running with other stuff under `gather_facts`), eg:

```suggestion
                    module.warn(f'mount_facts: ignoring repeat mounts for {mount_point}. "
```

Also (not a merge blocker), but it might be good to collapse to a single warning that lists them all to cut down on the boilerplate parts.

s-hertel replied:
I committed your suggestion, but I will add an update to collapse it to a single warning.

s-hertel replied:
I ran into an issue with symlinked sources, so if it's not a blocker, I'm going to leave this as-is. For example /etc/mtab is symlinked to /proc/mounts and since the module collects from both, if I include the source in the warning, the same warning for the same mount points is given twice. Here's what I was attempting:

```python
    duplicates_msg = ""
    # FIXME: would need to deduplicate by os.path.realpath(src)
    for src, mnts in file_source_mount_points.items():
        if len(set(mnts)) == len(mnts):
            continue
        duplicates = [mnt for mnt in mnts if mnts.count(mnt) > 1]
        duplicates_msg += f"{src}: {', '.join(duplicates)}\n"

    if module.params["include_aggregate_mounts"] is None and duplicates_msg:
        module.warn(f"mount_facts: ignoring repeat mounts for the following sources with duplicates:\n{duplicates_msg}"
                    "You can disable this warning by configuring the 'include_aggregate_mounts' option as True or False.")

```

s-hertel replied:
I collapsed this to a single warning in the latest commit, and fixed the symlinks issue so the aggregate_mounts don't contain duplicates.

nitzmahone commented on lib/ansible/modules/mount_facts.py:
```diff
@@ -0,0 +1,637 @@
+  include_aggregate_mounts:
+    description:
+      - Whether or not the module should return the C(aggregate_mounts) list in C(ansible_facts).
+      - When this is V(null), a warning will be emitted multiple mounts for the same mount point are found.
```
```suggestion
      - When this is V(null), a warning will be emitted if multiple mounts for the same mount point are found.
```

nitzmahone commented on lib/ansible/modules/mount_facts.py:
```diff
@@ -0,0 +1,637 @@
+    type: float
+  on_timeout:
+    description:
+      - The action to take when listing mounts or mount information exceeds O(timeout).
```
```suggestion
      - The action to take when gathering mount information exceeds O(timeout).
```

nitzmahone commented on lib/ansible/modules/mount_facts.py:
```diff
@@ -0,0 +1,637 @@
+    default: mount
+  timeout:
+    description:
+      - This is the maximum number of seconds to query for each mount. When this is V(null), wait indefinitely.
```
```suggestion
      - This is the maximum number of seconds to wait for each mount to complete. When this is V(null), wait indefinitely.
```

nitzmahone commented on lib/ansible/modules/mount_facts.py:
```diff
@@ -0,0 +1,637 @@
+  timeout:
+    description:
+      - This is the maximum number of seconds to query for each mount. When this is V(null), wait indefinitely.
+      - Configure in conjunction with O(on_timeout) to try to skip unresponsive mounts.
```
```suggestion
      - Configure in conjunction with O(on_timeout) to skip unresponsive mounts.
```

nitzmahone commented on lib/ansible/modules/mount_facts.py:
```diff
@@ -0,0 +1,637 @@
+        supports_check_mode=True,
+    )
+    if (seconds := module.params["timeout"]) is not None and seconds <= 0:
+        module.fail_json(msg="argument 'timeout' must be a positive integer or null, not {seconds}")
```
(since the argspec says `float`):
```suggestion
        module.fail_json(msg="argument 'timeout' must be a positive number or null, not {seconds}")
```

nitzmahone commented on lib/ansible/modules/mount_facts.py:
```diff
@@ -0,0 +1,637 @@
+    default: ~
+    type: list
+    elements: str
+  mount_binary:
```
IIUC the way this works, the mount binary wouldn't be able to accept args, since the value of `mount_binary` is also the "name" of its associated source. I'm wondering if it would make more sense to always use a fixed key like `mount` or `mount_binary` to refer to that in sources- then `mount_binary` could include args if needed?

s-hertel replied:
I could change sources to include a fixed key for the mount_binary instead, but mount_binary is looked up using `get_bin_path` so args aren't supported for that reason. Should I just run the mount_binary as a command instead of looking up the binary? I was hesitant to support args, since it can impact the output format (like dragonflybsd has a -p option to format like fstab instead), which would not match the supported regexes for mount output.

s-hertel replied:
I updated sources to use the fixed key "mount" in the latest commit. I haven't added a way to pass args to the mount_binary.

nitzmahone replied:
WFM- that at least opens more possibilities for passing args in the future- agreed you wouldn't want to pass anything that might change the format, but there are a crapton of args that can filter/expand what it can see and otherwise change things that could potentially be useful/needed at some point.

nitzmahone reviewed (commented):
Yep, I personally like this UX a lot better. A few minor suggested tweaks, and one last "oh, maybe it should..." change that might be a little bigger, but I think we're pretty much there. Thanks!

s-hertel committed:
Apply suggestions from code review

Co-authored-by: Matt Davis <6775756+nitzmahone@users.noreply.github.com>

s-hertel committed:
Fix syntax error

s-hertel committed:
fix layered mounts warning so it's only given once

determine source uniqueness with the realpath so aggregate_mounts doesn't contain repeats

test duplicate sources

revise how sources are handled so mount is only used as a fallback by
default again

s-hertel committed:
fix linux only test

mattclay commented on lib/ansible/modules/mount_facts.py:
```diff
@@ -0,0 +1,651 @@
+            stdout = run_mount_bin(module, module.params["mount_binary"])
+            results = [(source, *astuple(mount_info)) for mount_info in gen_mounts_from_stdout(stdout)]
+        else:
+            seen.add(os.path.realpath(source))
```
If different symlinks refer to the same source, then the `source in seen` check above won't detect the duplicate. Checking the real source here would protect against that:

```suggestion
            if (real_source := os.path.realpath(source)) in seen:
                continue
                
            seen.add(real_source)
```

mattclay replied:
I just realized the change I suggested doesn't resolve the issue I pointed out, since it would result in skipping a path that isn't a symlink. The seen check needs to be done in one place, or a separate set used for real paths.

s-hertel replied:
I have a warning above if non-symlink duplicates are given, so I think skipping those too is correct. How does https://github.com/ansible/ansible/pull/83508/commits/ef1f410f02f0a78de307cb9f403791b9241ebb55 look?

s-hertel committed:
Fix logic for deduplicating symlinks

mattclay commented on lib/ansible/modules/mount_facts.py:
```diff
@@ -0,0 +1,650 @@
+
+    seen = set()
+    for source in sources:
+        if (real_source := os.path.realpath(source)) in seen:
```
This does have the side-effect of turning `mount` into a path even though it's not. For the purposes of this check it should be fine, unless one of the actual sources happens to be a `mount` file in the current directory.

s-hertel replied:
Oh, good catch. let me fix that

s-hertel committed:
mark mount as seen rather than a path

nitzmahone reviewed (approved):
Yep, this resolves the UX concerns I had- thanks!
